### DISCUSSION TITLE
Possible bug docker-in-docker? could not connect to docker daemon

### DISCUSSION THREAD

**Discussion by mandrasch** [community/community#63776] [Codespaces] [answered]
### Select Topic Area

Bug

### Body

Hey, 

I'm experimenting with docker-in-docker and codespaces.

The [`.devcontainer.json`](https://github.com/mandrasch/ddev-craftcms-vite/blob/main/.devcontainer/devcontainer.json): 

```
{
  "image": "mcr.microsoft.com/devcontainers/universal:2",
  "features": {
    "ghcr.io/devcontainers/features/docker-in-docker:2": {},
    "ghcr.io/ddev/ddev/install-ddev:latest": {}
  },
```

1 times out of 15 I get the following error, it seems random:

```
2023-08-16 19:07:58.801Z: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```

Only a full rebuild solves it or starting a completely new codespace.

Update 16.08.2023: Full error log here Full output: https://gist.github.com/mandrasch/3b001bccdb8ab9ab75ebd7ddc6973727

```
time="2023-08-16T19:07:55.475998600Z" level=error msg="containerd did not exit successfully" error="exit status 1" module=libcontainerd
failed to start daemon: error while opening volume store metadata database: timeout
x\"..." type=io.containerd.runtime.v1
```

Update 16.08.2023: Issue found, missed it first time https://github.com/devcontainers/features/issues/634

Repo: https://github.com/mandrasch/ddev-craftcms-vite

Thanks in advance for any hints, much appreciated!

Best regards,
Matthias

**Comment by nathan-hello** [comment]
in all of my .devcontainer.json files, i add a mount for the .sock file in the host machine to the one in the docker. This way any docker commands i run is actually being ran on the host machine. this also means any devcontainer + docker shenanigans i run into, it's because of the host machine and not the devcontainer, which is easier to control (though kind of not with codespaces but you get what i mean). anyways, here's the mount. 

```json
    "mounts": [
        "type=bind,source=/var/run/docker.sock,target=/var/run/docker.sock"
    ],
```

**Reply by mandrasch** [reply]
thanks for very much reply! 

I have no idea right now if that would be compatible with the DDEV integration. But I'll try to explore this topic.

**Reply by rfay** [reply]
Use `docker context ls` to see the docker contexts available. This doesn't really have anything to do with DDEV. You just need:
1. The `docker` client to be in your $PATH
2. The appropriate context to be selected (`docker context use <context>`)
3. `docker ps` to work.

When those things are working, DDEV will work. The standard Codespaces setup for DDEV does this by itself. https://github.com/ddev/ddev/blob/master/.devcontainer/devcontainer.json - but so does most every other Codespaces setup.

**Reply by mandrasch** [reply]
Thanks very much for clarification, @rfay!

I added these commands as debug output into my demo repo (https://github.com/mandrasch/ddev-craftcms-vite/blob/main/.devcontainer/postCreateCommand.sh#L5).

Tried 10 times yesterday with new codespace instances - the "docker connect" error did not happen... 🤷 Hard to reproduce unfortunately.

Hope to catch one of these errors in future and provide the output here.

**Reply by mandrasch** [reply]
Oh well, catched it right after posting at run number 15! 

```bash
2023-08-16 19:07:57.280Z: chmod +x ./.devcontainer/postCreateCommand.sh && ./.devcontainer/postCreateCommand.sh
2023-08-16 19:07:57.389Z: + docker context ls
2023-08-16 19:07:58.675Z: NAME        DESCRIPTION                               DOCKER ENDPOINT               KUBERNETES ENDPOINT   ORCHESTRATOR
default *   Current DOCKER_HOST based configuration   unix:///var/run/docker.sock                         swarm
```

`echo $PATH`:
``` /home/codespace/.local/bin:/home/codespace/.dotnet:/home/codespace/nvm/current/bin:/home/codespace/.php/current/bin:/home/codespace/.python/current/bin:/home/codespace/java/current/bin:/home/codespace/.ruby/current/bin:/home/codespace/.local/bin:/usr/local/python/current/bin:/usr/local/py-utils/bin:/usr/local/oryx:/usr/local/go/bin:/go/bin:/usr/local/sdkman/bin:/usr/local/sdkman/candidates/java/current/bin:/usr/local/sdkman/candidates/gradle/current/bin:/usr/local/sdkman/candidates/maven/current/bin:/usr/local/sdkman/candidates/ant/current/bin:/usr/local/rvm/gems/default/bin:/usr/local/rvm/gems/default@global/bin:/usr/local/rvm/rubies/default/bin:/usr/local/share/rbenv/bin:/opt/conda/bin:/usr/local/php/current/bin:/usr/local/nvs:/usr/local/share/nvm/current/bin:/usr/local/hugo/bin:/usr/local/dotnet/current:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/rvm/bin
2023-08-16 19:07:58.682Z: /home/codespace/.local/bin:/home/codespace/.dotnet:/home/codespace/nvm/current/bin:/home/codespace/.php/current/bin:/home/codespace/.python/current/bin:/home/codespace/java/current/bin:/home/codespace/.ruby/current/bin:/home/codespace/.local/bin:/usr/local/python/current/bin:/usr/local/py-utils/bin:/usr/local/oryx:/usr/local/go/bin:/go/bin:/usr/local/sdkman/bin:/usr/local/sdkman/candidates/java/current/bin:/usr/local/sdkman/candidates/gradle/current/bin:/usr/local/sdkman/candidates/maven/current/bin:/usr/local/sdkman/candidates/ant/current/bin:/usr/local/rvm/gems/default/bin:/usr/local/rvm/gems/default@global/bin:/usr/local/rvm/rubies/default/bin:/usr/local/share/rbenv/bin:/opt/conda/bin:/usr/local/php/current/bin:/usr/local/nvs:/usr/local/share/nvm/current/bin:/usr/local/hugo/bin:/usr/local/dotnet/current:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/rvm/bin
```

```
+ docker ps
2023-08-16 19:07:58.801Z: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
2023-08-16 19:07:58.871Z: postCreateCommand failed with exit code 1. Skipping any further user-provided commands.
```

Full output: https://gist.github.com/mandrasch/3b001bccdb8ab9ab75ebd7ddc6973727

In terminal I also get 

```
@mandrasch ➜ /workspaces/ddev-craftcms-vite (main) $ docker ps
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```

**Reply by rfay** [reply]
I wonder if it's a timing thing, docker daemon just isn't ready yet. Why don't you try this approach, https://github.com/ddev/ddev/blob/f04f31c2273665cf0649088ed65db0d1c48e6e20/.buildkite/test.sh#L49-L59

**Reply by mandrasch** [reply]
Also wondered if it is about timing.

But shouldn't `docker ps` work afterwards in the codespaces terminal if it's timing related and postCreateCommand is executed too early? 🤔

Still get this in the "broken" codespace instance after the failed creation:   

```
@mandrasch ➜ /workspaces/ddev-craftcms-vite (main) $ docker ps
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```

**Reply by rfay** [reply]
If you put that loop in there waiting for `docker ps` to work, and it does work there, then yes it should work afterward. If it doesn't you likely have the wrong docker context. It looks like in your current output you only have one docker context, so seems like that would be hard to do.

**Reply by mandrasch** [reply]
I added the output of `cat /tmp/dockerd.log`: https://gist.github.com/mandrasch/3b001bccdb8ab9ab75ebd7ddc6973727#file-dockerd-log

**Reply by rfay** [reply]
Looks like it was a mess on the previous run (if I understand right):
> failed to start daemon: error while opening volume store metadata database: timeout

I guess there are a few questions here. What are you doing differently than the [DDEV-provided devcontainer.json](https://github.com/ddev/ddev/blob/56afe4bee4170688435510c79135d32d63b7d824/.devcontainer/devcontainer.json#L2-L14) that doesn't seem to have ever given any trouble?

**Reply by mandrasch** [reply]
The image used is different of course, but got it from [DDEV docs](https://ddev.readthedocs.io/en/stable/users/install/ddev-installation/#github-codespaces) + it's the default one for codespaces now: https://hub.docker.com/_/microsoft-devcontainers-universal
It's based on Ubuntu.

[`docker-in-docker`](https://github.com/devcontainers/features/tree/main/src/docker-in-docker) states that any debian based image should be fine, therefore ubuntu should be okay. 🤔 

But the image used in [DDEV-provided devcontainer.json](https://github.com/ddev/ddev/blob/56afe4bee4170688435510c79135d32d63b7d824/.devcontainer/devcontainer.json#L2-L14) is Debian 11 bullseye, maybe worth a shot.

I also use `forwardPorts:`. Otherwise I did not change much:

```
{
  "image": "mcr.microsoft.com/devcontainers/universal:2",
  "features": {
    "ghcr.io/devcontainers/features/docker-in-docker:2": {},
    "ghcr.io/ddev/ddev/install-ddev:latest": {}
  },
  // https://code.visualstudio.com/docs/devcontainers/containers#_always-forwarding-a-port
  "forwardPorts": [5174],
  // these are just labels for codespaces ports tab, these will not create ports by default
  "portsAttributes": {
    "3306": {
      "label": "database"
    },
    "8027": {
      "label": "mailhog"
    },
    "8036": {
      "label": "phpmyadmin"
    },
    "8080": {
      "label": "web http"
    },
    "8443": {
      "label": "web https"
    },
    // our custom vite port for ddev + codespaces
    "5174":{
      "label": "vite"
    }
  },

  "postCreateCommand": "chmod +x ./.devcontainer/postCreateCommand.sh && ./.devcontainer/postCreateCommand.sh"
}
```

**Reply by mandrasch** [reply]
I could also open an issue at https://github.com/devcontainers/features. Guess the whole thing would be reproducable with some luck and without install-ddev feature.

**Reply by mandrasch** [reply]
Also saw "* Docker-in-docker: Add retries until docker daemon starts" in changelog from docker-in-docker `2.3.0`, but this version was already used in my broken instance. https://github.com/devcontainers/features/releases/tag/feature_docker-in-docker_2.3.0

https://github.com/devcontainers/features/pull/637

**Reply by rfay** [reply]
Hmm, one thing that you (and DDEV) should know is that docker is provided with no fiddling in Codespaces. With no devcontainer.json at all, you get docker running by default when opening any repo.

**Reply by mandrasch** [reply]
Thanks very much, @rfay! 

Since other devs experience it in the issue below as well, at least we can confirm it has nothing to do with adding another feature like ddev. 

https://github.com/devcontainers/features/issues/634

Unfortunately I just found the issue now, did not find it when I first searched through the issues for github.com/devcontainers/features/ ... Must have missed it.

**Reply by rfay** [reply]
Congrats! Looks like that will be in next release, and other people have the problem.

**Reply by mandrasch** [reply]
👍  The fix was published in docker-in-docker `2.3.0`, which was already used in my broken example from today. But hopefully there will be another patch in future.

**Reply by rfay** [reply]
It does claim that, but I guess you should chime up in the issue and explain that it's not fixed :(

**Comment by othman922** [comment]
#

**Comment by samruddhikhandale** [comment, marked as answer]
Hi 👋 

Responded on the https://github.com/devcontainers/features/issues/634 issue with more details.

> Hi @mandrasch 👋 
> 
> In your dev container, docker is added by the `universal` image. See https://github.com/devcontainers/images/blob/main/src/universal/.devcontainer/devcontainer.json#L64-L66. 
> 
> Hence, adding the `docker-in-docker` Feature has no impact as the docker-init script is  old. See https://gist.github.com/mandrasch/3b001bccdb8ab9ab75ebd7ddc6973727#file-creation-log-txt-L168. (I think this behavior should be updated.)
> 
> The prod image was built with Feature version 2.2.1 (released on Aug 3rd) which does not contain the retry logic. I'll work on releasing a new `universal` image which would be built with v2.3.1. 
> 
> Let me know if that makes sense.
> 
> > I'll work on releasing a new universal image which would be built with v2.3.1.
> 
> Opened https://github.com/devcontainers/images/pull/705. In the meanwhile, @mandrasch can you use the dev image? (`mcr.microsoft.com/devcontainers/universal:dev`). Let me know if that helps! 
> 
> Also, can you remove the `docker-in-docker` Feature from your dev container as that one is unnecessary (given that the universal image already installs it)?

**Reply by mandrasch** [reply]
Just for documentation:

For all people interested in this, I also responded in the issue: https://github.com/devcontainers/features/issues/634#issuecomment-1681609252
